nexus/src/app/background/plan_execution.rs

```diff
@@ -0,0 +1,327 @@
+
+use super::common::BackgroundTask;
+use anyhow::Context;
+use futures::future::BoxFuture;
+use futures::stream;
+use futures::FutureExt;
+use futures::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_types::blueprint::{Blueprint, OmicronZonesConfig};
+use serde_json::json;
+use slog::Logger;
+use std::collections::BTreeMap;
+use std::net::SocketAddr;
+use std::sync::Arc;
+use tokio::sync::watch;
+use uuid::Uuid;
+
+/// Background task that takes a [`Blueprint`] and realizes the change to
+/// the state of the system based on the `Blueprint`.
+pub struct PlanExecutor {
+    rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+}
+
+impl PlanExecutor {
+    // Temporary until we wire up the background task
+    #[allow(unused)]
+    pub fn new(
+        rx_blueprint: watch::Receiver<Option<Arc<Blueprint>>>,
+    ) -> PlanExecutor {
+        PlanExecutor { rx_blueprint }
+    }
+}
+
+impl BackgroundTask for PlanExecutor {
+    fn activate<'a, 'b, 'c>(
+        &'a mut self,
+        opctx: &'b OpContext,
+    ) -> BoxFuture<'c, serde_json::Value>
+    where
+        'a: 'c,
+        'b: 'c,
+    {
+        async {
+            // Get the latest blueprint, cloning to prevent holding a read lock
+            // on the watch.
+            let blueprint = self.rx_blueprint.borrow().clone();
+
+            let Some(blueprint) = blueprint else {
+                warn!(&opctx.log,
+                      "Plan execution: skipped";
+                      "reason" => "no blueprint");
+                return json!({"error": "no blueprint" });
+            };
+
+            let log =
+                opctx.log.new(o!("description" => blueprint.description()));
+
+            let result = realize_blueprint(&log, &blueprint).await;
+
+            // Return the result as a `serde_json::Value`
+            match result {
+                Ok(()) => json!({}),
+                Err(errors) => {
+                    let errors: Vec<_> = errors
+                        .into_iter()
+                        .map(|e| format!("{:#}", e))
+                        .collect();
+                    json!({"errors": errors})
+                }
+            }
+        }
+        .boxed()
+    }
+}
+
+async fn realize_blueprint(
+    log: &Logger,
+    blueprint: &std::sync::Arc<Blueprint>,
+) -> Result<(), Vec<anyhow::Error>> {
+    match &**blueprint {
+        Blueprint::OmicronZones(zones) => deploy_zones(log, zones).await,
+    }
+}
+
+async fn deploy_zones(
+    log: &Logger,
+    zones: &BTreeMap<Uuid, (SocketAddr, OmicronZonesConfig)>,
+) -> Result<(), Vec<anyhow::Error>> {
+    let errors: Vec<_> = stream::iter(zones.clone())
+        .filter_map(|(sled_id, (addr, config))| async move {
+            let client = sled_agent_client::Client::new(
+                &format!("http://{}", addr),
+                log.clone(),
+            );
+            let result =
+                client.omicron_zones_put(&config).await.with_context(|| {
```

Comment:
For my understanding -- what are the semantics of `omicron_zones_put`? Will it remove any zones that are currently running but not in the new config?

Reply:
Correct. `omicron_zones_put` will create a zone if one doesn't already exist in the set, and remove any that are not in the new config.